### PR TITLE
prov/rxm + tcp: Infrastructure to support rxm passthru to tcp

### PR DESCRIPTION
**Pull request by shefty**
With a few changes to the tcp provider, we can eliminate most of the overhead of the rxm provider, including the rxm protocol.  RxM is still used to multiplex data over multiple tcp connections.  The tcp provider handles all messaging and protocol.  All data transfer operations fall directly to the tcp provider after the corresponding connection is selected.

This series is not complete.  Work is needed to hook into the fi_getinfo path.  And changes are needed to the tcp tag matching capability to match based on the source address.  (The current implementation of relying using a connection to identify the source only works after we've established a connection.  However, a receive can be posted prior to the connection being setup).

Posting for early review before the patch build-up grows even longer.

### CODE REVIEW

File: prov/tcp/src/tcpx_shared_ctx.c [hunk 1]
```diff
@@ -305,15 +317,25 @@ static int tcpx_srx_close(struct fid *fid)
...
 		xfer_entry = container_of(entry, struct tcpx_xfer_entry, entry);
+		if (srx->cq) {
+			tcpx_cq_report_error(&srx->cq->util_cq, xfer_entry,
+					      FI_ECANCELED);
+		}
 		ofi_buf_free(xfer_entry);
 	}
 
+	if (srx->cq)
+		ofi_atomic_inc32(&srx->cq->util_cq.ref);
```
Comment (ooststep) [inline prov/tcp/src/tcpx_shared_ctx.c:355]: should this be `ofi_atomic_dec32`?

Comment (shefty) [reply prov/tcp/src/tcpx_shared_ctx.c:355]: most definitely

**Comment by shefty** [comment]
I forgot I had a PR open for this.  Anyway, the full set of changes now pass fabtests, including ubertest.  Pass thru support is enabled automatically based on fi_getinfo.  I plan to add a follow up patch to allow disabling it as a fallback option.

This series removes rxm from data transfer operations.  However, rxm still maintains a cq separate from that used by tcp.  I'm not sure we can merge the cq's and still support the API generically.

Once the changes pass CI, we need to analyze the performance.

**Comment by shefty** [comment]
Failing Intel CI tests -- likely because counter support is not there, and a lot of CI tests check for it.

**Comment by shefty** [comment]
Updated to include counter support.  When passthru counters are enabled, counter wait support is disabled.

**Comment by shefty** [comment]
Failures are related to counter support, which apparently isn't quite right yet.

**Comment by shefty** [comment]
Counter support was updated.  Intel CI failures is with psm2 getinfo tests.

**Comment by shefty** [comment]
@ooststep @j-xiong - This is a pile of changes, but should be ready for review and possible merging sometime after the v1.14 release.

File: prov/tcp/src/tcpx_shared_ctx.c [hunk 2]
```diff
@@ -310,6 +310,53 @@ tcpx_match_tag_addr(struct tcpx_rx_ctx *srx, struct tcpx_ep *ep, uint64_t tag)
...
 
+static bool
+tcpx_srx_cancel_rx(struct tcpx_rx_ctx *srx, struct slist *queue, void *context)
+{
+	struct slist_entry *cur, *prev;
+	struct tcpx_xfer_entry *xfer_entry;
+
+	assert(fastlock_held(&srx->lock));
+
+	slist_foreach(queue, cur, prev) {
```
Comment (j-xiong) [inline prov/tcp/src/tcpx_shared_ctx.c:321]: Is this macro safe when slist_remove() is involved?

File: prov/rxm/src/rxm_ep.c [hunk 3]
```diff
@@ -1138,7 +1138,7 @@ static int rxm_ep_msg_cq_open(struct rxm_ep *rxm_ep)
 		return ret;
 	}
 
-	if (cq_attr.wait_obj != FI_WAIT_FD)
+	if ((cq_attr.wait_obj != FI_WAIT_FD))// || domain->passthru)
```
Comment (j-xiong) [inline prov/rxm/src/rxm_ep.c:1141]: The domain->passthru part is commented out. Does it need to be uncommented or removed?

File: prov/rxm/src/rxm_rma.c [hunk 4]
```diff
@@ -514,3 +514,219 @@ struct fi_ops_rma rxm_ops_rma = {
...
+
+	ret = rxm_get_conn(ep, src_addr, &conn);
+	if (ret)
+		goto unlock;
+
+	ret = fi_read(conn->msg_ep, buf, len, desc, src_addr, addr,
+		      key, context);
+unlock:
+	ofi_ep_lock_release(&ep->util_ep);
+	return ret;}
```
[the remark this answers is not in the document]
Comment (shefty) [reply prov/rxm/src/rxm_rma.c:539]: oops - updated in next push

File: prov/tcp/src/tcpx_shared_ctx.c (hunk 2 above)
Comment (shefty) [reply prov/tcp/src/tcpx_shared_ctx.c:321]: If we continued with the iteration, no, but we exit after matching.

File: prov/rxm/src/rxm_ep.c (hunk 3 above)
Comment (shefty) [reply prov/rxm/src/rxm_ep.c:1141]: This was a debug hack and will be removed in the next push.

File: prov/rxm/src/rxm_ep.c [hunk 5]
```diff
@@ -1138,7 +1138,7 @@ static int rxm_ep_msg_cq_open(struct rxm_ep *rxm_ep)
 		return ret;
 	}
 
-	if (cq_attr.wait_obj != FI_WAIT_FD)
+	if ((cq_attr.wait_obj != FI_WAIT_FD))
```
Comment (j-xiong) [inline prov/rxm/src/rxm_ep.c:1141]: nitpick: don't need double parentheses.

**Review by j-xiong — APPROVED** [review]

File: prov/rxm/src/rxm_ep.c (hunk 5 above)
Comment (shefty) [reply prov/rxm/src/rxm_ep.c:1141]: Bad cleanup from removing the debug code I had there from before.  Will update

**Comment by shefty** [comment]
We are analyzing the performance impact of these changes.  So far, the results are inconclusive  It'll take some time to figure out why we don't see improvements basically everywhere.  But in some cases, the results are much worse.  Will update last commit to disable by default.  I don't want to carry these many changes out of tree, but until we can analyze the performance, we'll want to keep them disabled.

**Pull request merged**